### DISCUSSION TITLE
Awful documentation

### DISCUSSION THREAD

**Discussion by adnan-smlatic** [valhalla/valhalla#4495] [Ideas]

**Comment by kevinkreiser** [comment]
> I do not understand how amazing projects such as this are created, and then end up with minimal surface level documentation that's practically useless for anyone new?

> It seems like an absolutely lovely project and the things that I have managed to get working are great, but for everything else it's just a fucking pain in the ass. Half the things aren't even written and you have to figure it out yourself, the other half is written across 70 sentences out of context in 6 different pages and you have to connect it yourself. God forbid you come across an error or something not working as expected because you won't find anything on it.

if you do not understand why free software that people volunteer their time to create isnt perfectly documented and doesnt do everything you want perfectly, then i would guess there could be two explanations for that. either you are very new to software development OR you feel entitled inspite of reality.. maybe both?

> I would absolutely love to contribute to this project and use it, but I think the starting point should be with the people that are experienced in it to present in a way where that would be possible in the first place.

from my perspective, this first contribution of yours has been quite enough already. perhaps you might work on humility and tact before engaging in other open source projects.

now some general comments; yes we know the documentation is inadequate in many ways. our time is not infinite, many things we do are in various stages of "iteration." we can't write bug-free code every time we set our fingers to keyboards despite ample testing. this project concerns itself with a pretty complex set of topics, which require a high time investment and often deep subject knowledge. not all problems are trivial so we need to prioritize our time among many competing needs. for me, id say my attention is usually: impact first (new feature, performance) followed by testing/bug fixes and then usually last is documentation.

in my experience it is unlikely that a person will want to commit the investment to become a routing expert and stick with it for years. the barrier to entry is high, not just because we dont write tutorials for you about it but because that is the nature of the problem space. could we do more work to educate prospective contributors? of course we could. is it worth the time investment? i am highly skeptical!

i have mentored around 20 or more people on exactly this topic and this project and i can tell you from that experience that if their lifes mission wasnt already to become a routing expert then they will not stick with the project. most people are here for the utility of it (their project or their jobs dictate it) and thats it. the people who come for the love of it, they will put in the work to figure it out and ask questions etc. it is those people that we want contributing to the project. i'll take a handful of them over fair-weather users any day of the week.

**Reply by adnan-smlatic** [reply]
![this-bear-can-smell-retards](https://github.com/valhalla/valhalla/assets/79746744/a38219ba-b5cc-4bc7-8693-312979e38cbf)

**Reply by nilsnolde** [reply]
I find it quite offensive to identify Kevin as a bear

**Reply by michaelkirk** [reply]
I'm sorry that you have to receive this kind of abuse.

Do what you like of course, but please don't feel like you *owe* any response at all to a belligerent stranger on the internet, who's not willing to put minimum effort into being neither clear nor constructive.

**Reply by nilsnolde** [reply]
I find it rather funny tbh. It’s self destructive, it’s what makes it fun:)

**Reply by kevinkreiser** [reply]
I personally had a good laugh from it too hahaha 

I knew i didn't have to reply but I had a free min and it seemed like a bit of harmless retort
